### PR TITLE
Improve criteria performance

### PR DESCRIPTION
**Pull request by alyst**
Another pass on optimizing criterion performance:
  * replace `Ones() * M` and similar constructs with `sum(M, dims=...)` -- it should be faster
  * replace `tr(A' * B)` with `dot(A, B)` where possible -- should be much faster
  * more uses of `∇Q` (if available) as a temporary matrix for intermediate operations to reduce memory footprint
  * simplify and unify the criterion code for CF and Oblimin (now it is more apparent that it's a weighted sum of column and row sums)
  * add Equamax and Parsimax on top of the generalized CF/Oblimin criterion code
  * TargetRotation: skip weighting if all target matrix elements are specified

### CODE REVIEW

File: src/methods/geomin.jl [hunk 1]
```diff
@@ -16,16 +16,16 @@ end
...
-    part = exp.((log.(Λsq) * v) ./ k)
-    Q = u' * part
+    part = exp.(sum(log.(Λsq), dims=2) ./ k)
+    Q = sum(part)
     if !isnothing(∇Q)
-        mul!(∇Q, part, v')
-        ∇Q .*= (2 / k) .* inv.(Λsq) .* Λ
+        # ∇Q === Λsq
+        part .*= (2 / k)
+        ∇Q .= Λ ./ Λsq .* part
```
Comment (alyst) [inline src/methods/geomin.jl:28]:
```suggestion
        ∇Q .= (2 / k) .* Λ ./ Λsq .* part
```

File: src/methods/oblimin.jl [hunk 2]
```diff
@@ -1,27 +1,27 @@
...
 rotation methods given a value for `gamma`:
 
-- `gamma = 0` → [`Quartimax`](@ref)
-- `gamma = 0.5` → [`Biquartimax`](@ref)
-- `gamma = 1` → [`Varimax`](@ref)
-- `gamma = p/2` → Equamax
+- *γ = 0*: [`Quartimax`](@ref)
+- *γ = 1/2*: [`Biquartimax`](@ref)
+- *γ = 1*: [`Varimax`](@ref)
+- *γ = p/2*: [`Equamax`](@ref)
```
Comment (p-gw) [inline src/methods/oblimin.jl:19]: Can we add the equivalency for `Parsimax` as well?

File: src/utils.jl [hunk 3]
```diff
@@ -50,6 +50,31 @@ function centercols!(m::AbstractMatrix)
     return m
 end
 
+# generic function for calculating the
```
Comment (p-gw) [inline src/utils.jl:53]: use proper docstrings here

File: src/utils.jl [hunk 4]
```diff
@@ -50,6 +50,31 @@ function centercols!(m::AbstractMatrix)
...
+# criterion and the gradient of the method
+# based on the weighted sums of columns and rows
+function weighted_sums_criterion_and_gradient!(
```
Comment (p-gw) [inline src/utils.jl:56]: I think `weighted_sums_criterion_and_gradient!` should live in src/methods/methods.jl instead.

**Review by p-gw — CHANGES_REQUESTED** [review]
Thanks, looks very good. We can merge this after the minor changes I proposed. 

Also, can you add `Parsimax` and `Equamax` to the `rotation_methods` docs for completeness? 

closes #27; closes #28

**Comment by alyst** [comment]
@p-gw Thank you for the review. I've addressed your comments + tweaked the infomax and biquartimin.

**Review by p-gw — APPROVED** [review]
Thanks, very nice PR!

**Pull request merged**